reuse serverAuthCode obtained on login

p-shapovalov opened the pull request:
## Description
Reuse serverAuthCode obtained on login in flutter/plugins/pull/4179

## Related issues
flutter/flutter#57712
flutter/flutter#57741

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

## Breaking Change
Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is not a breaking change.

stuartmorgan commented:
Thanks for the contribution!

> * [x]  I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.

Same question here as in the other PR; I'm not seeing these tests in the PR.

stuartmorgan commented:
I've fixed the title, but this was not the case:

> * [x]  The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`

Please keep in mind for future submissions that the checklist is to help make sure all the important steps are followed; just checking all the boxes without doing the steps causes PRs to take longer to triage and review.

p-shapovalov commented:
fixed

p-shapovalov commented:
Just in case clarification here too. I've removed serverAuthCode from getToken response because that was wrong - android/ios native response doesn't contain serverAuthCode field

And I've added serverAuthCode to user data response according to my changes because serverAuthCode obtained on that step

p-shapovalov commented:
Thank you!

cyanglaz commented:
> Post-submit LGTM (with one question), since this should have had two approvals before landing.

Is it a new policy that we are enforcing for the whole repo? Or we are doing this for certain types of changes?

jainkrati commented:
Thanks for the fix. Somehow I am still getting null serverAuthCode. Is this change already pushed in latest published package for google_sign_in?

Also, are you aware of any ongoing or past fix for ios for this scenario? I am integrating google sign in for my app and need both these fixes. Will appreciate any help :)

cc @p-shapovalov  @stuartmorgan

p-shapovalov commented:
You need to specify correct server client id to get serverAuthCode. Please ensure you have it in SERVER_CLIENT_ID(iOS) and default_web_client_id(Android). For android you also can pass value directly into plugin constructor

jainkrati commented:
Thanks for the quick tip @p-shapovalov . I made the changes but it keeps crashing everytime google sign in opens in ios after account selection. Strangely, the exception is not throw in the console but the device keeps disconnecting. 

I believe its a config issue but I am not able to put a finger on it. So sharing the details here, please let me know if you notice anything that needs to change in the way I am using/configuring the plugin:
1. In GoogleService-Info.plist, following values are present:
<key>CLIENT_ID</key>
	<string>IOSClientID.apps.googleusercontent.com</string>
	<key>REVERSED_CLIENT_ID</key>
	<string>com.googleusercontent.apps.IOSClientID</string>
	<key>SERVER_CLIENT_ID</key>
	<string>webClientID.apps.googleusercontent.com</string>
	<key>PLIST_VERSION</key>
	<string>1</string>
	<key>BUNDLE_ID</key>
	<string>myAppsBundleID</string> 
2. I am instantiating googleSignIn this way:
 account = await GoogleSignIn(clientId: IOSClientID, hostedDomain: "", scopes: [
          'email',
          'https://www.googleapis.com/auth/contacts.readonly'
        ]).signIn();
3. If account is not null then accessing serverAuthCode:
 if (account != null) {
        final GoogleSignInAuthentication googleSignInAuthentication =
            await account.authentication;
        print("serverAuthCode is ${googleSignInAuthentication.serverAuthCode}");
}

p-shapovalov commented:
I'm not sure what exactly wrong with your configuration, because I didn't use google sign in for ios, I guess you can try to catch exception somewhere in https://github.com/flutter/plugins/blob/master/packages/google_sign_in/google_sign_in/lib/google_sign_in.dart#L376 

However according to your sample
> If account is not null then accessing serverAuthCode:

Please check deprecation warning and use account.serverAuthCode instead

stuartmorgan commented:
Please do not use PRs for support requests.

tylerstephens814 commented:
@jainkrati I'm experiencing the same issue.

tylerstephens814 commented:
@jainkrati I initially had the same problem as you. (My implementation is very similar to yours) Then I realized that my `GoogleService-Info.plist` in VSCode did not match in Xcode. Fixed that and everything starting working.

jainkrati commented:
Thanks @tylerstephens814 :)

Yesterday even I got google sign in working on ios by fixing two issues:
1. Adding SERVER_CLIENT_ID 
2. Removing and re-adding all entries of plist file from XCode instead of VS Code. Apparently there were some extra characters when we added from VS Code last time.
